feat: add automatic timestamps for various tables

iPurpl3x opened the pull request:
<!--
IMPORTANT: We applaud pull requests, they excite us every single time. As we have an obligation to maintain a healthy code standard and quality, we take sufficient time for reviews. Please do create a separate pull request per change/issue/feature; we will ask you to split bundled pull requests.
-->

**Changes proposed in this pull request:**
Database migrations that add timestamps to various tables: 
- `groups`
- `group_user`
- `group_permission`
- `tags`
- `discussion_tag`
- `post_mentions_post`
- `post_mentions_user`

The timestamps are not really used anywhere in Flarum, but can be very handy for analytics purposes.

This PR is kind of a follow-up of https://github.com/flarum/likes/pull/28

**Reviewers should focus on:**
The migrations run without error and don't cause any unwanted side-effects (including performance aspects).

**Necessity**

- [x] Has the problem that is being solved here been clearly explained?
- [x] If applicable, have various options for solving this problem been considered?
- [x] For core PRs, does this need to be in core, or could it be in an extension?
- [x] Are we willing to maintain this for years / potentially forever?

**Confirmed**

- [ ] Frontend changes: tested on a local Flarum installation.
- [x] Backend changes: tests are green (run `composer test`).
- [x] Core developer confirmed locally this works as intended.
- [x] Tests have been added, or are not appropriate here.

**Required changes:**

- [ ] Related documentation PR: (Remove if irrelevant)
- [ ] Related core extension PRs: (Remove if irrelevant)

iPurpl3x committed:
Add automatic timestamps for various tables:
- groups
- group_user
- group_permission
- tags
- discussion_tag
- post_mentions_post
- post_mentions_user

askvortsov1 reviewed (approved):
Haven't tested locally, but makes sense!

SychO9 reviewed (commented):
Could you please for every timestamp add it to the relevant model's `$dates` array property for proper casting?

also very minor but there seems to be a space in the filename:
`2022_05_20_000005_add_created_at_to_ post_mentions_post_table.php`

davwheat commented:
The only issue with this is that it would set `created_at` to the time the migration is run, which isn't accurate and could pollute statistics.

Would it not be better to set to null initially, then update the default value to be the current time?

iPurpl3x committed:
Make timestamps nullable

iPurpl3x committed:
Fix whitespace in file name

iPurpl3x committed:
Add casts to models for the newly added timestamps

iPurpl3x commented:
@SychO9 Sure, I added the casts (for the tables that have corresponding models) and fixed the file name (good that you saw that one!)

@davwheat I have thought of that. The thing is that:

_IF_ we choose to update the value of the column automatically in the DB (and not use Laravel, which might slightly impact performance) then the `DEFAULT` is going to be `current_timestamp()` so I would need to run a second migration that sets all the old values to `NULL` after the column(s) have been added. Do you want me to do that ?

davwheat commented:
In my opinion, running two migrations to do that (can they be merged into two steps of one migration? I have no idea) would be the best option.

I'd be interested to hear the opinions of other maintainers though.

iPurpl3x commented:
@davwheat If I don't use the `\Flarum\Database\Migration` helper, and write the migration by hand, I might be able to merge it all into one file. 

I also would like to hear opinions of others...

SychO9 commented:
How about creating the columns first, then modifying them to use current (and use current on update where appropriate). That avoids having to modify all rows.

```php
use Illuminate\Database\Schema\Blueprint;
use Illuminate\Database\Schema\Builder;

return [
    'up' => function (Builder $schema) {
        $schema->table('groups', function (Blueprint $table) {
            $table->timestamp('created_at')->nullable();
        });

        // do this manually because dbal doesn't recognize timestamp columns
        $connection = $schema->getConnection();
        $prefix = $connection->getTablePrefix();
        $connection->statement("ALTER TABLE $prefix`groups` MODIFY created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP");
    },

    'down' => function (Builder $schema) {
        $schema->table('groups', function (Blueprint $table) {
            $table->dropColumn('created_at');
        });
    }
];
```

Also it might be good to have a comment on each migration that mentions the columns were added for third-party analytical purposes.

iPurpl3x commented:
Ah yes, that's pretty clever @SychO9, I will try that I keep you updated.

Agreed for the comments.

iPurpl3x committed:
Update timespamp migrations so that the existing rows stay `NULL`

iPurpl3x commented:
I have now updated all the migrations, using the idea of @SychO9 to makes that all existing rows in the DB stay with NULL on the timestamps and don't get a false value of the time when the migration is done.

Hopefully that makes this PR now mergeable 🤞🏻

davwheat reviewed (approved):
Tested locally. Seems to work as intended on non-prefixed DBs.

![image](https://user-images.githubusercontent.com/7406822/171531736-515f0e52-34c7-4390-bbd4-550da36324b6.png)
![image](https://user-images.githubusercontent.com/7406822/171531742-e2ffdd3e-325e-4317-b72d-470552feb008.png)

davwheat commented on extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
As per tests failing, should this (and others) instead be....

```suggestion
        $connection->statement("ALTER TABLE `${prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

SychO9 replied:
yup, my bad

```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

davwheat replied:
I'll go through and update them -- no need to inconvenience Rafael with that :)

davwheat replied:
Oh yeah, I don't have permission to push to GB's fork! 🙈 

The suggestions are all there for @iPurpl3x to bulk-commit anyway :)

iPurpl3x replied:
Done @davwheat

davwheat commented on extensions/mentions/migrations/2022_05_20_000005_add_created_at_to_post_mentions_post_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_post` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

davwheat commented on extensions/mentions/migrations/2022_05_20_000006_add_created_at_to_post_mentions_user_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`post_mentions_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}post_mentions_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

davwheat commented on extensions/tags/migrations/2022_05_20_000003_add_timestamps_to_tags_table.php:
```diff
@@ -0,0 +1,33 @@
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`tags` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
+        $connection->statement("ALTER TABLE $prefix`tags` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}tags` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
        $connection->statement("ALTER TABLE `{$prefix}tags` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```

davwheat commented on extensions/tags/migrations/2022_05_20_000004_add_created_at_to_discussion_tag_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`discussion_tag` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}discussion_tag` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

davwheat commented on framework/core/migrations/2022_05_20_000000_add_timestamps_to_groups_table.php:
```diff
@@ -0,0 +1,33 @@
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`groups` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
+        $connection->statement("ALTER TABLE $prefix`groups` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}groups` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
        $connection->statement("ALTER TABLE `{$prefix}groups` MODIFY updated_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP ON UPDATE CURRENT_TIMESTAMP");
```

davwheat commented on framework/core/migrations/2022_05_20_000001_add_created_at_to_group_user_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`group_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}group_user` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

davwheat commented on framework/core/migrations/2022_05_20_000002_add_created_at_to_group_permission_table.php:
```diff
@@ -0,0 +1,30 @@
+        // do this manually because dbal doesn't recognize timestamp columns
+        $connection = $schema->getConnection();
+        $prefix = $connection->getTablePrefix();
+        $connection->statement("ALTER TABLE $prefix`group_permission` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```
```suggestion
        $connection->statement("ALTER TABLE `{$prefix}group_permission` MODIFY created_at TIMESTAMP NULL DEFAULT CURRENT_TIMESTAMP");
```

iPurpl3x committed:
Fix string interpolation
